[WIP] Add existential structs to Checked C

abeln opened the pull request:
Design doc: https://github.com/microsoft/checkedc/wiki/%5BProposal%5D-Add-Existential-Structs-to-CheckedC

PR with tests: https://github.com/microsoft/checkedc/pull/382

I'll update the PR to be more meaningful when we're closing to merging this.

Current state of the PR: there are lots of TODOs, so it'd be helpful to get feedback both on the current state of things and prioritization of issues.

abeln commented on include/clang/AST/Type.h:
```diff
@@ -4414,8 +4414,10 @@ class TypeVariableType : public Type, public llvm::FoldingSetNode {
-};
 
+  // TODO: document
+  static const int ExistentialVariableDepthStart = 10000;
```
This is no longer needed and I'll remove it.

abeln replied:
Done

abeln commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -7372,6 +7404,175 @@ void Parser::ParseCheckedPointerSpecifiers(DeclSpec &DS) {
+  if (ExpectAndConsume(tok::l_paren)) return;
+
+  if (Tok.getKind() != tok::identifier) {
+    // TODO: add proper error message
```
This will be fixed before merging.

abeln replied:
Done

dtarditi commented on include/clang/AST/Type.h:
```diff
@@ -4441,6 +4440,57 @@ class TypedefType : public Type {
+  void Profile(llvm::FoldingSetNodeID &ID) {
+    Profile(ID, TypeVar, InnerType);
+  }
+  static void Profile(llvm::FoldingSetNodeID &ID, const Type *TypeVar, QualType InnerType) {
```
I believe you could 'profile' existential types by using:
```
ID.AddPointer(TypeVar)
ID.AddPointer(InnerType.getAsOpaquePtr()).
```
This would let you remove the caching code that you've added in ASTContext.cpp, and use FoldingSets as used for other types.   You'd have to move the explanatory comment about the unique type objects not necessarily being canonical elsewhere.

abeln replied:
As discussed, will leave as a TODO (copied your comment to https://github.com/microsoft/checkedc-clang/issues/661).

dtarditi commented on include/clang/AST/RecursiveASTVisitor.h:
```diff
@@ -1263,6 +1268,7 @@ DEF_TRAVERSE_TYPELOC(FunctionProtoType, {
 DEF_TRAVERSE_TYPELOC(UnresolvedUsingType, {})
 DEF_TRAVERSE_TYPELOC(TypedefType, {})
 DEF_TRAVERSE_TYPELOC(TypeVariableType, {})
+DEF_TRAVERSE_TYPELOC(ExistentialType, {}) // TODO: is this correct?
```
I believe this code needs to traverse the type locations for the type variable and the the body of the existing type, which are in the TypeLoc for the ExistentialType.

abeln replied:
Unfortunately, I still don't understand how type locations work. For example, this method (TraverseTypeLoc) sometimes calls itself recursively, but sometimes it calls TraverseType or even TraverseStmt. Maybe we can chat about this code tomorrow?

dtarditi commented on include/clang/Sema/DeclSpec.h:
```diff
@@ -623,6 +631,19 @@ class DeclSpec {
+  bool isUnpackSpecified() const { return Unpack_specified; }
+  SourceLocation getUnpackSpecLoc() const { return UnpackLoc; }
+
+  // TODO: add comment about how this method is also used for generics (the comment for existentials is below).
```
Yes, these comments need to be cleaned up.

abeln replied:
Done

dtarditi commented on lib/AST/ASTContext.cpp:
```diff
@@ -2087,6 +2087,11 @@ TypeInfo ASTContext::getTypeInfoImpl(const Type *T) const {
     break;
 
+  case Type::Existential:
+    // '_Exists(T, struct Foo<T>)' gets erased to struct 'Foo<T>', so calculate
```
Nit: the term 'erasure' may be unfamiliar to future compiler developers.  I suggest you change the phrase 'gets erased to struct 'Foo<T>' to 'is represented at runtime as 'struct Foo<T>'.  You could note parenthetically that this is an erasure strategy and that the use of T is restricted so that it doesn't affect runtime layout)

abeln replied:
Done

dtarditi commented on lib/AST/ASTContext.cpp:
```diff
@@ -9263,6 +9270,9 @@ QualType ASTContext::mergeTypes(QualType LHS, QualType RHS,
            "Equivalent type variable types should have already been handled!");
     return {};
+  case Type::Existential:
+    // TODO: is this correct?
```
I am not sure this is correct.  I'll look at this in my next review.

dtarditi commented on lib/AST/ASTStructuralEquivalence.cpp:
```diff
@@ -819,6 +819,13 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
+  case Type::Existential: {
+    const ExistentialType *Exist1 = cast<ExistentialType>(T1);
+    const ExistentialType *Exist2 = cast<ExistentialType>(T2);
+    // TODO: rename bounds variables according to depth. 
```
I'll look at this in my next review.

dtarditi commented on lib/Sema/SemaType.cpp:
```diff
@@ -1565,6 +1565,15 @@ static QualType ConvertDeclSpecToType(TypeProcessingState &state) {
   }
+  case DeclSpec::TST_exists: {
+    auto InnerType = S.GetTypeFromParser(DS.getRepAsType());
+    assert(DS.typeVariables().size() == 1 && "Expected exactly one type variable for an existential");
```
Is this guaranteed?  I don't believe it is.  If it isn't, you shouldn't assert this, as it could cause a compiler crash.  Instead, there should be a diagnostic that this isn't supported yet.  You can use setInvalidType to invalidate the type.

abeln replied:
Yes, that there is just one type variable is enforced by the parser.

dtarditi commented on lib/AST/ExprClassification.cpp:
```diff
@@ -445,6 +445,8 @@ static Cl::Kinds ClassifyInternal(ASTContext &Ctx, const Expr *E) {
   case Expr::RangeBoundsExprClass:
     llvm_unreachable("should not classify bounds expressions");
+  case Expr::PackExprClass:
+    return Cl::CL_PRValue;
```
It would be good to add a comment about this.

abeln replied:
Done

dtarditi commented on lib/AST/Type.cpp:
```diff
@@ -3631,6 +3631,9 @@ static CachedProperties computeCachedProperties(const Type *T) {
     return Cache::get(cast<PipeType>(T)->getElementType());
   case Type::TypeVariable:
     return CachedProperties(ExternalLinkage, false);
+  case Type::Existential:
```
I believe you should use the cached properties of the body of the existential type.   This should be treated the same way as other constructed types such as array types and function types are constructed.

abeln replied:
Done, but how can we test this?

dtarditi replied:
This code still isn't looking at the body of the existential type.  I believe you would need to define a variable with an existential type in one compilation unit, declare it another, and make sure that they link.

abeln replied:
Done. Added a TODO for the test.

dtarditi commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2406,10 +2406,24 @@ Decl *Parser::ParseDeclarationAfterDeclaratorAndAttributes(
+        if (D.getDeclSpec().isUnpackSpecified()) {
+          // TODO: make more robust, add asserts, etc.
+          auto TpeVars = D.getDeclSpec().typeVariables();
+          assert(TpeVars.size() == 1 && "Expected exactly one type variable in _Unpack specifier");
```
Please issue a diagnostic instead of asserting.  This will crash the compiler if the user has more than 1 type variable.

abeln replied:
Done. Added a test as well.

dtarditi commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2406,10 +2406,24 @@ Decl *Parser::ParseDeclarationAfterDeclaratorAndAttributes(
+          auto TypeArg = TypeArgument { TVar, SourceInfo };
+          auto *InitExpr = Init.get();
+          // TODO: typing the lhs might fail: need to handle gracefully.
+          auto InitType = cast<ExistentialType>(InitExpr->getType().getTypePtr())->innerType();
```
This needs to be cleaned up too, as it could a compiler crash.

abeln replied:
Done

dtarditi commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3557,6 +3559,57 @@ ExprResult Parser::ParseReturnValueExpression() {
+    return ExprError();
+  }
+
+  auto ExistTpe = ParseTypeName();
```
Why use the variable name ExistTpe instead of ExistType?

abeln replied:
Fixed

dtarditi commented on lib/Sema/DeclSpec.cpp:
```diff
@@ -1051,6 +1053,13 @@ bool DeclSpec::setSpecItypeforany(SourceLocation Loc, const char *&PrevSpec, uns
 }
 
+bool DeclSpec::setUnpackSpec(SourceLocation Loc, const char *&PrevSpec, unsigned &DiagID) {
+  // TODO: add error handling
```
It would be good to add error handling now so that we don't have to revisit this file later.

abeln replied:
I can't think of any error conditions that apply. I think the comment was just wrong.

abeln commented on include/clang/AST/Expr.h:
```diff
@@ -3139,6 +3139,64 @@ class CompoundLiteralExpr : public Expr {
+/// Used like so '_Exists(T, struct Foo<T>) foo = _Pack(fooImpl, _Exists(T, struct Foo<T>), int)'.
+/// A pack expression is a triple (expr, exist_typ, subst), indicating that we type
+/// 'expr' with the existential type 'exist_typ', using the substitution 'T -> subst', where
+/// 'T' is the type variable introduced by the existential.
```
Done

abeln commented on include/clang/AST/Expr.h:
```diff
@@ -3139,6 +3139,64 @@ class CompoundLiteralExpr : public Expr {
+  /// The expression that is being packed.
+  Expr *PackedExpr = nullptr;
+
+  /// The result type of the pack operation.
```
Done

abeln commented on include/clang/AST/Expr.h:
```diff
@@ -3139,6 +3139,64 @@ class CompoundLiteralExpr : public Expr {
+  QualType getExistentialType() const { return ExistTpe; }
+
+  /// Return the "substitution" type that is used to verify that
+  /// the underlying expr can be packed with the existential type.
```
Done

abeln commented on lib/AST/StmtProfile.cpp:
```diff
@@ -1134,6 +1134,11 @@ void StmtProfiler::VisitCastExpr(const CastExpr *S) {
 }
 
+void StmtProfiler::VisitPackExpr(const PackExpr *S) {
+  // TODO: implement
```
Done

abeln commented on lib/AST/StmtPrinter.cpp:
```diff
@@ -1692,6 +1692,13 @@ void StmtPrinter::VisitInteropTypeExpr(
 }
 
+void StmtPrinter::VisitPackExpr(PackExpr *Node) {
+  // TODO: print the rest of the components
```
Done

abeln commented on lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -202,6 +202,12 @@ class AggExprEmitter : public StmtVisitor<AggExprEmitter> {
+
+  // Checked C
+
+  /// To codegen a pack expression such as '_Pack(foo-expr, _Exists(T, struct Foo<T>), subst-type)',
```
Done

abeln commented on lib/CodeGen/CodeGenTypes.cpp:
```diff
@@ -386,8 +386,12 @@ llvm::Type *CodeGenTypes::ConvertType(QualType T) {
-    return ConvertRecordDeclType(RT->getDecl());
+  if (const RecordType *RT = dyn_cast<RecordType>(Ty)) {
+    auto *RecDecl = RT->getDecl();
+    // TODO: explain this step
```
Done

abeln commented on lib/CodeGen/CodeGenTypes.cpp:
```diff
@@ -555,7 +559,13 @@ llvm::Type *CodeGenTypes::ConvertType(QualType T) {
+  case Type::Existential: {
+    // Existential types desugar into their inner types.
+    // e.g.: '_Exists(T, struct Foo<T>)' becomes 'struct Foo<T>', which in turn
+    // becomes 'struct Foo', where uses of 'T' are replaces by 'void'.
```
Done

abeln commented on lib/Sema/SemaExpr.cpp:
```diff
@@ -14367,6 +14369,27 @@ void Sema::SetDeferredBoundsCallBack(void *OpaqueData, ParseDeferredBoundsCallBa
+  // Example: suppose the existential is '_Exists(T, struct Foo<T>)'. Then the witness should have
+  // type 'struct Foo<int>'.
+  auto *Exist = dyn_cast<ExistentialType>(ExistTpe.getTypePtr());
+  assert(Exist && "Expected existential type in pack expression");
```
Added a check to the caller of this function. Now if this fails, it really should be an assertion error.

abeln commented on lib/Serialization/ASTReader.cpp:
```diff
@@ -6630,6 +6630,11 @@ void TypeLocReader::VisitTypeVariableTypeLoc(TypeVariableTypeLoc TL) {
 }
 
+void TypeLocReader::VisitExistentialTypeLoc(ExistentialTypeLoc TL) {
+  // TODO: implement
```
Done

abeln commented on lib/Serialization/ASTWriterStmt.cpp:
```diff
@@ -850,6 +850,11 @@ void ASTStmtWriter::VisitBoundsCastExpr(BoundsCastExpr *E) {
 }
 
+void ASTStmtWriter::VisitPackExpr(PackExpr *E) {
+  // TODO: implement
```
Done

abeln commented on lib/Serialization/ASTWriter.cpp:
```diff
@@ -364,6 +364,11 @@ void ASTTypeWriter::VisitTypeVariableType(const TypeVariableType *T) {
 }
 
+void ASTTypeWriter::VisitExistentialType(const ExistentialType *TL) {
+  // TODO: implement
```
Done

abeln commented:
@dtarditi I've addressed most of your comments. PTAL. I've added a couple of questions.
I also updated the companion PR with more tests: https://github.com/microsoft/checkedc/pull/382

I'm tracking the TODOs in the master issues for existentials (https://github.com/microsoft/checkedc-clang/issues/661) and generics (https://github.com/microsoft/checkedc-clang/issues/644).

dtarditi commented on lib/Sema/CheckedCSubst.cpp:
```diff
@@ -478,3 +487,245 @@ QualType Sema::SubstituteTypeArgs(QualType QT, ArrayRef<TypeArgument> TypeArgs)
+
+  /// For an existential type, add the introduced variable to the set of bound variables
+  /// and recurse on the inner type.
+  QualType TransformExistentialType(TypeLocBuilder &TLB, ExistentialTypeLoc TL) {
```
I believe you should be able to override this to get the behavior you want.

dtarditi commented on lib/Sema/DeclSpec.cpp:
```diff
@@ -1051,6 +1053,12 @@ bool DeclSpec::setSpecItypeforany(SourceLocation Loc, const char *&PrevSpec, uns
   return false;
 }
 
+bool DeclSpec::setUnpackSpec(SourceLocation Loc, const char *&PrevSpec, unsigned &DiagID) {
```
Does this need error checking? If not, please add a comment as to why not.  I am concerned about combining this with other specifiers.

dtarditi commented on lib/Sema/TreeTransform.h:
```diff
@@ -5815,6 +5815,26 @@ QualType TreeTransform<Derived>::TransformTypeVariableType(TypeLocBuilder &TLB,
+  QualType NewInnerType = getDerived().TransformType(ExistType->innerType());
+  if (NewInnerType.isNull()) return QualType();
+  auto RawRes = SemaRef.ActOnExistentialType(SemaRef.Context, NewTypeVar.getTypePtr(), NewInnerType);
+  if (!RawRes) llvm_unreachable("Could not build existential type");
```
You should return QualType() isntead of asserting here.

abeln replied:
Done

dtarditi commented on lib/Sema/TreeTransform.h:
```diff
@@ -5815,6 +5815,26 @@ QualType TreeTransform<Derived>::TransformTypeVariableType(TypeLocBuilder &TLB,
+  // TODO: the proper way to do the rebuilding is to have a RebuildExistentialType method. Add one?
+  auto *ExistType = TL.getTypePtr();
+  auto Quals = TL.getType().getQualifiers().getAsOpaqueValue();
+  auto TypeVar = QualType(ExistType->typeVar(), 0 /* Quals */);
```
The comment is confusing because you just defined Quals.

dtarditi commented on lib/Sema/TreeTransform.h:
```diff
@@ -5815,6 +5815,26 @@ QualType TreeTransform<Derived>::TransformTypeVariableType(TypeLocBuilder &TLB,
+  auto *ExistType = TL.getTypePtr();
+  auto Quals = TL.getType().getQualifiers().getAsOpaqueValue();
+  auto TypeVar = QualType(ExistType->typeVar(), 0 /* Quals */);
+  QualType NewTypeVar = getDerived().TransformType(TypeVar);
```
You should call the version of TransFormType that takes a TypeLocBuilder.  See for example TransformVectorType().

dtarditi commented on lib/Sema/TreeTransform.h:
```diff
@@ -5815,6 +5815,26 @@ QualType TreeTransform<Derived>::TransformTypeVariableType(TypeLocBuilder &TLB,
+  if (NewTypeVar.isNull()) return QualType();
+  QualType NewInnerType = getDerived().TransformType(ExistType->innerType());
+  if (NewInnerType.isNull()) return QualType();
+  auto RawRes = SemaRef.ActOnExistentialType(SemaRef.Context, NewTypeVar.getTypePtr(), NewInnerType);
```
You are missing the logic to avoid rebuilding here.   Most functions have the logic "getDerved(). AlwaysRebuild() || check that types are different".

dtarditi commented on lib/Sema/TreeTransform.h:
```diff
@@ -12667,6 +12687,12 @@ TreeTransform<Derived>::TransformRangeBoundsExpr(RangeBoundsExpr *E) {
                                              E->getRParenLoc());
 }
 
+template<typename Derived>
```
There needs to be code to transform the body and types of the pack expression.

dtarditi commented on lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -628,6 +628,11 @@ llvm::DIType *CGDebugInfo::CreateType(const TypeVariableType *TvT) {
+llvm::DIType *CGDebugInfo::CreateType(const ExistentialType *Ty) {
+  // TODO: is this correct?
+  return nullptr;
+}
```
I believe you should call CreateType on the body of the existential.  Please file an issue to "try out debugging of code with existential types."

abeln replied:
Done. See https://github.com/microsoft/checkedc-clang/issues/661#issuecomment-533344812

dtarditi commented on lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -2910,6 +2915,9 @@ llvm::DIType *CGDebugInfo::CreateTypeNode(QualType Ty, llvm::DIFile *Unit) {
     return CreateType(cast<BuiltinType>(Ty));
   case Type::TypeVariable:
     return CreateType(cast<TypeVariableType>(Ty));
+  case Type::Existential:
```
Yes, this specific line is correct :).

dtarditi commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2406,10 +2406,35 @@ Decl *Parser::ParseDeclarationAfterDeclaratorAndAttributes(
+          auto TypeArg = TypeArgument { TVar, SourceInfo };
+          // The initializer must have an existential type.
+          // e.g. _Unpack (T) struct Foo<T> fooAbs = expr // expr must have type '_Exists(A, struct Foo<A>)'
+          auto InitExprType = InitExpr->getType().getCanonicalType().getTypePtr();
```
You should use the getType->getAs<ExistentialType>() instead.  In general, if you are manipulating CanonicalTypes directly in semantic checking, that is worrisome.

abeln replied:
Good point. Done.

abeln commented:
@dtarditi the build succeeds now
https://dev.azure.com/msresearch/CheckedC/_build/results?buildId=342238

abeln commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3557,6 +3559,60 @@ ExprResult Parser::ParseReturnValueExpression() {
+
+  TypeSourceInfo *ExistTInfo = nullptr;
+  auto UnwrappedExist = Actions.GetTypeFromParser(ExistType.get(), &ExistTInfo);
+  if (!ExistentialType::classof(UnwrappedExist.getCanonicalType().getTypePtr())) {
```
Great suggestion. Done.

abeln commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3557,6 +3559,60 @@ ExprResult Parser::ParseReturnValueExpression() {
+    return ExprError();
+  }
+
+  auto SubstTpe = ParseTypeName();
```
Fixed
